In [1]:
#Q2 part 1: 0/1 knapsack for stock portfolio
def stock_portfolio_knapsack(prices, returns, names, budget):
    n = len(prices)
    # Create DP table: (n+1) x (budget+1)
    dp = [[0 for _ in range(budget + 1)] for _ in range(n + 1)]

    # Build table
    for i in range(1, n + 1):
        for w in range(0, budget + 1):
            if prices[i-1] <= w:
                dp[i][w] = max(dp[i-1][w], returns[i-1] + dp[i-1][w - prices[i-1]])
            else:
                dp[i][w] = dp[i-1][w]

    # Find which stocks are included
    w = budget
    portfolio = []
    for i in range(n, 0, -1):
        if dp[i][w] != dp[i-1][w]:
            portfolio.append(names[i-1])
            w -= prices[i-1]

    total_invested = sum([prices[names.index(stock)] for stock in portfolio])
    expected_return_value = sum([returns[names.index(stock)] for stock in portfolio]) * total_invested

    return portfolio[::-1], expected_return_value, total_invested


if __name__ == "__main__":
    names = ["APP", "AMZ", "FBB", "GOLG", "IBN", "MASO", "MDN", "TSLA", "VO", "WLM"]
    prices = [90, 1200, 150, 800, 40, 380, 160, 290, 25, 90]
    returns = [0.10, 0.16, 0.12, 0.15, 0.08, 0.25, 0.18, 0.09, 0.11, 0.06]

    budget = int(input("Enter the budget (up to $2500): "))
    if budget > 2500:
        print("Budget cannot exceed $2500.")
    else:
        portfolio, expected_return, total_invested = stock_portfolio_knapsack(prices, returns, names, budget)
        print(f"Portfolio: {portfolio}")
        print(f"Expected maximum return: ${expected_return:.2f}")
        print(f"Total amount invested: ${total_invested}")


Portfolio: ['APP', 'AMZ', 'FBB', 'IBN', 'MASO', 'MDN', 'TSLA', 'VO', 'WLM']
Expected maximum return: $2788.75
Total amount invested: $2425


***Q1 part 1- Output ***
Portfolio: ['APP', 'AMZ', 'FBB', 'IBN', 'MASO', 'MDN', 'TSLA', 'VO', 'WLM']
Expected maximum return: $2788.75
Total amount invested: $2425


*** Q2 part 2- Unlimited Stocks Pseudocode ***
1. Create a list of stocks with (expected return per dollar) = (return ratio) / (price).
2. Sort all stocks in descending order based on (expected return per dollar).
3. Initialize:
   total_return = 0
   total_invested = 0
   portfolio = []
4. For each stock in sorted list:
   a. If current total_invested + price <= budget:
       - Buy as much stock as possible (even fractions)
       - Update total_return += (price * return ratio)
       - Update total_invested += price
   b. Else:
       - Buy partial stock with remaining budget.
       - Calculate fraction = (remaining budget) / (price)
       - Update total_return += (fraction * price * return ratio)
       - Update total_invested += (remaining budget)
       - Add stock to portfolio with fraction
       - Break loop (budget is exhausted)
5. Output portfolio, total_return, total_invested.
